src/jitlayers.cpp

```diff
@@ -0,0 +1,396 @@
+                                      const object::ObjectFile &debugObj,
+                                      const RuntimeDyld::LoadedObjectInfo &L);
+
+namespace {
+
+using namespace llvm;
+using namespace llvm::object;
+using namespace llvm::orc;
+
+/// Do the registration.
+void NotifyDebugger(jit_code_entry* JITCodeEntry) {
+  __jit_debug_descriptor.action_flag = JIT_REGISTER_FN;
+
+  // Insert this entry at the head of the list.
+  JITCodeEntry->prev_entry = nullptr;
+  jit_code_entry* NextEntry = __jit_debug_descriptor.first_entry;
+  JITCodeEntry->next_entry = NextEntry;
+  if (NextEntry) {
+    NextEntry->prev_entry = JITCodeEntry;
+  }
+  __jit_debug_descriptor.first_entry = JITCodeEntry;
+  __jit_debug_descriptor.relevant_entry = JITCodeEntry;
+  __jit_debug_register_code();
+}
+
+// --------------------------------------------------------------------------
+
+class DebugObjectRegistrar {
+private:
+    void NotifyGDB(OwningBinary<ObjectFile> &DebugObj) {
+      const char *Buffer = DebugObj.getBinary()->getMemoryBufferRef().getBufferStart();
+      size_t      Size = DebugObj.getBinary()->getMemoryBufferRef().getBufferSize();
+
+      assert(Buffer && "Attempt to register a null object with a debugger.");
+      jit_code_entry* JITCodeEntry = new jit_code_entry();
+
+      if (!JITCodeEntry) {
+        llvm::report_fatal_error(
+          "Allocation failed when registering a JIT entry!\n");
+      } else {
+        JITCodeEntry->symfile_addr = Buffer;
+        JITCodeEntry->symfile_size = Size;
+
+        NotifyDebugger(JITCodeEntry);
+      }
+    }
+
+    std::vector<OwningBinary<ObjectFile>> SavedObjects;
+    std::unique_ptr<JITEventListener> JuliaListener;
+
+public:
+    DebugObjectRegistrar() : JuliaListener(CreateJuliaJITEventListener()) {}
+
+    template <typename ObjSetT, typename LoadResult>
+    void operator()(ObjectLinkingLayerBase::ObjSetHandleT, const ObjSetT &Objects,
+                  const LoadResult &LOS) {
+        auto oit = Objects.begin();
+        auto lit = LOS.begin();
+        while (oit != Objects.end()) {
+            auto &Object = *oit;
+            auto &LO = *lit;
+
+            OwningBinary<ObjectFile> SavedObject = LO->getObjectForDebug(*Object);
+
+            // If the debug object is unavailable, save (a copy of) the original object
+            // for our backtraces
+            if (!SavedObject.getBinary()) {
+                // This is unfortunate, but there doesn't seem to be a way to take
+                // ownership of the original buffer
+                auto NewBuffer = MemoryBuffer::getMemBufferCopy(Object->getData(), Object->getFileName());
+                auto NewObj = ObjectFile::createObjectFile(NewBuffer->getMemBufferRef());
+                SavedObject = OwningBinary<ObjectFile>(std::move(*NewObj),std::move(NewBuffer));
+            }
+            else
+                NotifyGDB(SavedObject);
+
+            SavedObjects.push_back(std::move(SavedObject));
```

Comment:
i vaguely feel like you meant `SavedObjects` to be `std::vector<std::unique_ptr<OwningBinary<ObjectFile>>>` so that it was managing the object lifetime (although we don't actually implement deletion at this time)